tests/eosio.system_ram_tests.cpp

```diff
@@ -235,4 +235,152 @@ BOOST_FIXTURE_TEST_CASE( buy_ram_self, eosio_system_tester ) try {
+
+   // bob should not be able to transfer any gifted ram either
+   // --------------------------------------------------------
+   BOOST_REQUIRE_EQUAL(ramtransfer(bob, "gft"_n, 1, ""), error("assertion failure with message: insufficient quota"));
+
+   // bob should not be able to receive a ram gift from another account while holding gifted ram already
+   // ---------------------------------------------------------------------------------------------------
+   BOOST_REQUIRE_EQUAL(giftram("gifter"_n, bob, 1, ""),
+      error("assertion failure with message: A single RAM gifter is allowed at any one time per account, currently holding RAM gifted by: gft"));
+
+   // To return the gift, bob needs 3574 bytes. Let's buy 3000, so with the 1400 we'll have enough 
+   // to return the gift and also do a 1 byte ramtransfer
+   // ---------------------------------------------------------------------------------------------
+   BOOST_REQUIRE_EQUAL(buyrambytes(bob, bob, 3'000u), success());
+   BOOST_REQUIRE_EQUAL(ungiftram(bob, "gft"_n, ""), success());
+   BOOST_REQUIRE_EQUAL(ramtransfer(bob, "gft"_n, 1u, ""), success());
+
+   // Now that bob doesn't hold gifted RAM anymore, it can receive a RAM gift from another account
+   // --------------------------------------------------------------------------------------------
+   BOOST_REQUIRE_EQUAL(giftram("gifter"_n, bob, 1, ""), success());
+
+   // and it should be able to return it only to the correct account
+   // --------------------------------------------------------------
+   BOOST_REQUIRE_EQUAL(ungiftram(bob, "gft"_n, ""),
+                       error("assertion failure with message: Returning RAM to wrong gifter, should be: gifter"));
+   BOOST_REQUIRE_EQUAL(ungiftram(bob, "gifter"_n, ""), success());
+
+   // create another test account with a smaller initial gift
+   // -------------------------------------------------------
+   const account_name dan = "dan.gft"_n;
+   create_account_with_resources("dan.gft"_n, "gft"_n, 0, 2'000u);
+   transfer("eosio", "dan.gft", core_sym::from_string("1000.0000"));
+   stake_with_transfer("eosio", "dan.gft", net, cpu);
+
+   // Make sure we can gift more, and more than once
+   // ----------------------------------------------
+   BOOST_REQUIRE_EQUAL(giftram("gft"_n, dan, 1'000u, ""), success());
+   BOOST_REQUIRE_EQUAL(giftram("gft"_n, dan, 1'000u, ""), success());
+   BOOST_REQUIRE_EQUAL(2'000u + 1'000u + 1'000u, (uint64_t)get_total_stake(dan)["ram_bytes"].as_int64());
+
+   // Let's make sure that when gifter RAM is returned, it is unencumbered and can be sold.
+   // -------------------------------------------------------------------------------------
+   BOOST_REQUIRE_LE((uint64_t)get_total_stake(bob)["ram_bytes"].as_int64(), 5'000u); // make sure bob has <5000 ram bytes
+   BOOST_REQUIRE_EQUAL(buyrambytes(bob, bob, 10'000u), success());
+   BOOST_REQUIRE_EQUAL(giftram(bob, "gifter"_n, 10'000u, ""), success());
```

Comment:
I guess this is tapping into some of the 3000 bytes of RAM bob bought earlier since the `buyrambytes` for 10,000 bytes wouldn't have given bob exactly 10,000 bytes after accounting for the fee.

That's okay. But it is relying on the assumption that there was already excess unencumbered RAM held by bob prior to `buyrambytes` call in the line above.

Reply:
> I guess this is tapping into some of the 3000 bytes of RAM bob bought earlier since the `buyrambytes` for 10,000 bytes wouldn't have given bob exactly 10,000 bytes after accounting for the fee.

Actually, `buyrambytes` accounts for the fee to try to buy the requested amount, even though it may not get the number of bytes exactly right.

https://github.com/eosnetworkfoundation/eos-system-contracts/blob/20a8598f17c49890e53e37db0f6679102940b8a3/contracts/eosio.system/src/delegate_bandwidth.cpp#L30-L31